First Draft of Signal/Signaler Article for FF.com (#11153)
microsoft/FluidFramework: docs/content/docs/concepts/signals.md

aliases:
  - "/docs/advanced/signals/"
  - "/docs/deep/signals/"
draft: true
---

> When using DDSes, data is sequenced and stored within the Fluid container to achieve synchronized shared state. For scenarios that involve shared persisted data, DDSes provide an effective way to communicate data so that it is retained in the container. However, there are many scenarios where we need to data communicate that is short-lived, in which the ordering and storage of said information would be wasteful and unnecessary.

Reviewer: Is there a one-sentence concrete example of short-lived data that we could add to the end of this paragraph?

> When using DDSes, data is sequenced and stored within the Fluid container to achieve synchronized shared state. For scenarios that involve shared persisted data, DDSes provide an effective way to communicate data so that it is retained in the container. However, there are many scenarios where we need to data communicate that is short-lived, in which the ordering and storage of said information would be wasteful and unnecessary.

Reviewer 2: ```suggestion
When using DDSes, data is sequenced and stored within the Fluid container to achieve synchronized shared state. For scenarios that involve shared persisted data, DDSes provide an effective way to communicate data so that it is retained in the container. However, there are many scenarios where we need to data communicate that is short-lived, in which the ordering and storage of said information would be wasteful and unnecessary.
```

> When using DDSes, data is sequenced and stored within the Fluid container to achieve synchronized shared state. For scenarios that involve shared persisted data, DDSes provide an effective way to communicate data so that it is retained in the container. However, there are many scenarios where we need to data communicate that is short-lived, in which the ordering and storage of said information would be wasteful and unnecessary.

Reviewer 2: ```suggestion
When using DDSes, data is sequenced and stored within the Fluid container to achieve synchronized shared state. For scenarios that involve shared persisted data, DDSes provide an effective way to communicate data so that it is retained in the container. However, there could be many scenarios where we need to communicate data that is short-lived, in which the ordering and storage of said information would be wasteful and unnecessary.
```

Reviewer 2: Just tried to re-phrase the last sentence a bit better. And an example as Ricky suggested would be great here!

> In contrast, the data that is communicated via signals is not retained in the container, which makes it an appropriate channel for transmitting transient data. A good analogy to help conceptualize signals is that they can be interpreted as UDP protocol, whereas standard ops can be thought of as TCP protocol. Signals are not guaranteed to be ordered on delivery relative to other signals and ops, but they still provide a useful communication channel for impermanent and short-lived data.

Reviewer: The paragraph begins "In contrast ..." but the preceding paragraph already introduced the contrast. This paragraph is just continuing the idea in the last sentence of the preceding paragraph.

> In contrast, the data that is communicated via signals is not retained in the container, which makes it an appropriate channel for transmitting transient data. A good analogy to help conceptualize signals is that they can be interpreted as UDP protocol, whereas standard ops can be thought of as TCP protocol. Signals are not guaranteed to be ordered on delivery relative to other signals and ops, but they still provide a useful communication channel for impermanent and short-lived data.

Author: Looking at it now, I agree that the analogy could be confusing to people unfamiliar with networking protocols and I think it's removal would be justified. 

To add more context, this was the analogy used to help describe the project to me and I did find it to be helpful since I was already familiar with the two networking protocols. The analogous part of the comparison is that UDP and signals are more lossy, faster, not guaranteed delivery, simpler, and have that "fire-and-forget" usage. TCP and DDSes guarantee more and are therefore more complicated in their implementation and usage in general.

> In contrast, the data that is communicated via signals is not retained in the container, which makes it an appropriate channel for transmitting transient data. A good analogy to help conceptualize signals is that they can be interpreted as UDP protocol, whereas standard ops can be thought of as TCP protocol. Signals are not guaranteed to be ordered on delivery relative to other signals and ops, but they still provide a useful communication channel for impermanent and short-lived data.

Reviewer 2: Agreed! Lets keep this simple. The networking analogy can get confusing. One thing we are trying to do with docs is that we make it easy for all levels of developers be able to understand these concepts better. I think you already explained it well with words like lossy, guarantee etc.

## Why are signals useful?

> Signals provide a communication channel for sharing short-lived information that does not need to be persisted in the Fluid op stream.

Reviewer: Have we decided that we're going to use "op" as a term or "operation"? If the former, is there a definition of "op" that we can link to?

Author: Yeah, Matt mentioned something similar [here](https://github.com/microsoft/FluidFramework/pull/11153#discussion_r923860523). I think replacing "Fluid op stream" with "Fluid container" would be better.

Reviewer 2: > Have we decided that we're going to use "op" as a term or "operation"? If the former, is there a definition of "op" that we can link to?

Don't think we have a definition for op in FF.com. Don't remember seeing anything. Good item to add to our the future work list.

By sending signals, you avoid the storage and sequencing of data that will not be relevant or useful in the long-term.

Signals are the most appropriate data channel in many user presence scenarios, where each user has the responsibility of sharing their current presence state to other connected users. In these scenarios, current presence data is short-lived, past presence state is irrelevant, and the shared data is not persisted on disconnect.

## How can I use signals in Fluid?

> The `Signaler` DataObject can be used to send communications via signals in a Fluid application. `Signaler` allows clients to send signals to other connected clients and add/remove listeners for specified signal types.

Reviewer: If there's reference documentation for Signaler, this first use of the term should be a link to that reference.

# Signaler

## Creation
Just like with DDSes, you can include `Signaler` as a shared object you would like to load in your `FluidContainer` schema.

Here is a look at how you would go about loading `Signaler` as part of the initial objects of the container:

```typescript
const containerSchema: ContainerSchema = {
    initialObjects: {
        signaler: Signaler,
    },
};

const { container, services } = await client.createContainer(containerSchema);

> const signaler = container.initialObjects.signaler as Signaler;

Reviewer: Is the `as Signaler` really needed? Its data type was specified earlier.

```

`signaler` can then be directly used in your Fluid application!

For more information on using `ContainerSchema` to create objects please see [Data modeling](https://fluidframework.com/docs/build/data-modeling/).

## API
`Signaler` provides a few simple methods to send signals and add/remove listeners to specific signals as well:
- `submitSignal(signalName: string, payload?: Jsonable)` - Sends a signal with a payload to its connected listeners
- `onSignal(signalName: string, listener: SignalListener)` - Adds a listener for the specified signal. Similar behavior as EventEmitter's `on` method.

> - `offSignal(signalName: string, listener: SignalListener)` - Removes a listener for the specified signal. Similar behavior as EventEmitter's `off` method.

Reviewer: These are puzzling without at least a one-sentence description of what the data type SignalListener is. 

Also, is the `listener` parameter a handler? I suspect it is, in which case it should be called `handler`. A listener and a handler are not quite the same thing.

Author: I think the usage of the term listener stems from the [EventEmitter Node.js class](https://nodejs.org/api/events.html#emitteroneventname-listener) and it's `emitter.on(eventName, listener)` and `emitter.off(eventName, listener)` methods. In the documentation, they define listener as "an event handler function" or a "callback function", so I thought the usage be appropriate.

Reviewer: OK. If this term has become industry-standard, we can use it.

```typescript
private static readonly focusRequestType = "focusRequest";
```

```typescript
container.on("connected", () => {
    this.signaler.submitSignal(FocusTracker.focusRequestType);
});
```

The connected clients are listening to this focus request signal, and they respond with their current focus state:

```typescript
this.signaler.onSignal(FocusTracker.focusRequestType, () => {
    this.sendFocusSignal(document.hasFocus());
});
```
This pattern adds cost however, as it forces every connected client to generate a signal.  Consider whether your scenario can be satisfied by receiving the signals naturally over time instead of requesting the information up-front. The mouse tracking in [PresenceTracker](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker) is an example where a newly connecting client does not request current state. Since mouse movements are frequent, the newly connecting client can instead simply wait to receive other users' mouse positions on their next mousemove event.
### Grouping Signal Types

> Rather than submitting multiple signals in response to an event, it is more cost-effective to submit one combined signal for that event and listen to that single signal instead. For example, imagine an application using the `Signal Request` pattern where a newly connected client requests for the color, focus state, and currently selected object of every other connected client on the page. If you submit a signal for each type of data requested, it would look something like this:

Reviewer: Either:

```suggestion
Rather than submitting multiple signals in response to an event, it is more cost-effective to submit one combined signal for that event and listen to that single signal instead. For example, imagine an application using the `Signal Request` pattern where a newly connected client asks for the color, focus state, and currently selected object of every other connected client on the page. If you submit a signal for each type of data requested, it would look something like this:
```
or: 

```suggestion
Rather than submitting multiple signals in response to an event, it is more cost-effective to submit one combined signal for that event and listen to that single signal instead. For example, imagine an application using the `Signal Request` pattern where a newly connected client requests the color, focus state, and currently selected object of every other connected client on the page. If you submit a signal for each type of data requested, it would look something like this:
```

```typescript
container.on("connected", () => {

>     this.signaler.submitSignal("colorRequest");

Reviewer: Need to explain briefly where these signal types, like "colorRequest" are defined.

Author: They are defined at the same time they are submitted. "colorRequest" is the signal name being sent to indicate the client is requesting the color of other users.

Reviewer: @WillieHabimana I'm still puzzled. The "colorRequest" signal is sent to something. So that something must have a definition of what "colorRequest" means. Where is that definition?

Author: Hmmm well It's being sent to other listening clients that will respond with submitting a signal with their color. Would having something like this help:
```typescript
this.signaler.onSignal("colorRequest", (clientId, local, payload) => {
            const clientIdColorMap = this.colorMap.get(this.audience.getMyself().userId);
            const clientColor =  clientIdColorMap.get(clientId);
            this.submitSignal("colorSignalType", {color: clientColor});
        });
```
I think the code snippet was just to give an idea of the best practice of what to do when sending multiple signals at once. Do the code snippets make the concept more confusing? If so, I think they could be removed.

Reviewer: Yes. This additional snippet is just what I had in mind.

Author: Perfect, sounds good!

    this.signaler.submitSignal("focusRequest");
    this.signaler.submitSignal("currentlySelectedObjectRequest");
});
```

> Each of the N connected clients would then respond with 3 signals as well (3N signals total).  To bring this down to N signals total, we can group these requests into a single request that captures all the required information:

Reviewer: Italicize the number variable:

```suggestion
Each of the _N_ connected clients would then respond with 3 signals as well (3_N_ signals total).  To bring this down to _N_ signals total, we can group these requests into a single request that captures all the required information:
```

aliases:
  - "/docs/advanced/signals/"
  - "/docs/deep/signals/"
draft: true
---

> Signals are similar to ops, but are not stored; they're useful to implement features like user presence. They provide a simple way to communicate transient data to connected clients within a Fluid application.

Reviewer 3: Our documentation teaches customers about ops, but it's not really something they use directly (instead they'll be indirectly generating ops by interacting with a DDS).  Maybe prefer to contrast against DDS usage rather than ops directly.

> They are lossy and not guaranteed to be ordered on delivery relative to other signals and ops. A good analogy to help conceptualize signals is that they can be interpreted as UDP protocol, whereas standard ops can be thought of as TCP protocol.

Author: Cool, I will definitely try to reach out! I put this since this is how signals were described by Skyler in my project description. He also included in the [ADO Task](https://dev.azure.com/fluidframework/internal/_workitems/edit/1146) to let the customers know that they are not persisted nor guaranteed. But I will definitely try to get more clarity around what exactly is guaranteed and what isn't when a signal is submitted.

> They are lossy and not guaranteed to be ordered on delivery relative to other signals and ops. A good analogy to help conceptualize signals is that they can be interpreted as UDP protocol, whereas standard ops can be thought of as TCP protocol.

Author: I also looked into the IFluidDataStoreRuntime to see it's signal guarantees here: [@fluidframework/datastore-definitions](https://github.com/microsoft/FluidFramework/tree/main/packages/runtime/datastore-definitions). 

> Signals are not persisted, ordered, or guaranteed. If a client is behind, the op state can be behind the signal state. For this reason people usually stick the currentSeq on the signal, so other clients can wait to process if they are behind.

> ## Why are signals useful?

Reviewer 3: I think the audience for this article will likely be someone who knows how to use DDSs, but has not yet used signals.  So, this section might benefit from framing it as a comparison against storing the information in a DDS -- instead of just "signals don't do X" it can be "using a DDS does X, but signals don't".

E.g. "When you use a DDS to communicate data, that data is retained as part of the container.  For data that is transient this storage can be wasteful.  By contrast, data communicated via signals is not retained, making them a good fit for transient data."

Signals provide a transient data channel for sharing short-lived information that does not need to be persisted in the Fluid op stream.

By sending signals, you avoid the storage and sequencing of data that will not be relevant or useful in the long-term.

Signals are the most appropriate data channel in many user presence scenarios, where each user has the responsibility of sharing their current presence state to other connected users. In these scenarios, current presence data is short-lived, past presence state is irrelevant, and the shared data is not persisted on disconnect.

## How can I use signals in Fluid?

> The `Signaler` DataObject can be used to effectively integrate and send communications via signals in a Fluid application. `Signaler` allows clients to efficiently send signals to other connected clients and add/remove listeners for specified signal types.

Reviewer 3: ```suggestion
The `Signaler` DataObject can be used to send communications via signals in a Fluid application. `Signaler` allows clients to send signals to other connected clients and add/remove listeners for specified signal types.
```

# Signaler

## Usage

> User presence scenarios are well-suited for `Signaler`, as users are required to tell other users their own information and their past data is mostly irrelavant. Using `Signaler` over other distributed data structures in these scenarios is beneficial, as its usage does not result in the storage of data that is not useful in the long-term.

Reviewer 3: Probably keep this section ("Signaler" -> "Usage") scenario-agnostic as possible, and move anything specific to user presence down in the Common Patterns section

Author: Ok cool. Looking at it no, I feel like the "Usage" section is sort of redundant within the context of this article, as everything said is already mentioned in the "Why are signals useful?" section. I think jumping straight into the creation of Signaler after its preamble in the "How can I use signals in Fluid?" would be appropriate.

Here is a look at how you would go about loading `Signaler` as part of the initial objects of the container:

```typescript
const containerSchema: ContainerSchema = {
    initialObjects: {
        signaler: Signaler,
    },
};

const { container, services } = await client.createContainer(containerSchema);

const signaler = container.initialObjects.signaler as Signaler;
```

`signaler` can then be directly used in your Fluid application!

For more information on using `ContainerSchema` to create objects please see [Data modeling](https://fluidframework.com/docs/build/data-modeling/).

## API
`Signaler` provides a few simple methods to send signals and add/remove listeners to specific signals as well:
- `submitSignal(signalName: string, payload?: Jsonable)` - Sends a signal with a payload to its connected listeners
- `onSignal(signalName: string, listener: SignalListener)` - Adds a listener for the specified signal. Similar behavior as EventEmitter's `on` method.
- `offSignal(signalName: string, listener: SignalListener)` - Removes a listener for the specified signal. Similar behavior as EventEmitter's `off` method.

## Common Patterns
### Signal Request

> When a client joins a collaboration session, they may need to receive pertinent information immediately after connecting the container.  To support this, they can request a specific signal be sent to them from other connected clients within the application. For example, in the [PresenceTracker](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker) we define a "focusRequest" signal type that a newly joining client uses to request the focus-state of each currently connected client:

Reviewer 3: ```suggestion
When a client joins a collaboration session, they may need to receive information about the current state immediately after connecting the container.  To support this, they can request a specific signal be sent to them from other connected clients. For example, in the [PresenceTracker](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker) example we define a "focusRequest" signal type that a newly joining client uses to request the focus-state of each currently connected client:
```

```typescript
private static readonly focusRequestType = "focusRequest";
```

```typescript
container.on("connected", () => {
    this.signaler.submitSignal(FocusTracker.focusRequestType);
});
```

The connected clients are listening to this focus request signal, and they respond with their current focus state:

```typescript
this.signaler.onSignal(FocusTracker.focusRequestType, () => {
    this.sendFocusSignal(document.hasFocus());
});
```

> When there are a lot of connected clients, usage of this request pattern can lead to high signal costs incurred from large amounts of signals being submitted all at the same time. While this pattern is helpful when a client is in need of relevant information, to limit signal costs it would be beneficial to examine whether or not the requested data will be quickly avaiable from other events being listened to within the application. The mouse tracking in [PresenceTracker](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker) is an example where a newly connecting client is not required to request a signal to receive every current mouse position on the document. Since mouse movements are frequent, the newly connecting client can simply wait to recieve other users mouse positions on their mousemove events.

Reviewer 3: ```suggestion
This pattern adds cost however, as it forces every connected client to generate a signal.  Consider whether your scenario can be satisfied by receiving the signals naturally over time instead of requesting the information up-front. The mouse tracking in [PresenceTracker](https://github.com/microsoft/FluidFramework/tree/main/examples/data-objects/presence-tracker) is an example where a newly connecting client does not request current state. Since mouse movements are frequent, the newly connecting client can instead simply wait to receive other users' mouse positions on their next mousemove event.
```

### Grouping Signal Types

> Rather than submitting multiple signal types in response to one specific event, it is more cost-effective to create one seperate signal type for that particular event and listen to that single signal instead. For example, imagine an application using the `Signal Request` pattern where a newly connected client requests for the color, focus state, and currently selected object of every other connected client on the page. If you submit a signal for each type of data requested, it would look something like this:

Reviewer 3: ```suggestion
Rather than submitting multiple signals in response to an event, it is more cost-effective to submit one combined signal for that event and listen to that single signal instead. For example, imagine an application using the `Signal Request` pattern where a newly connected client requests for the color, focus state, and currently selected object of every other connected client on the page. If you submit a signal for each type of data requested, it would look something like this:
```

```typescript
container.on("connected", () => {
    this.signaler.submitSignal("colorRequest");
    this.signaler.submitSignal("focusRequest");
    this.signaler.submitSignal("currentlySelectedObjectRequest");
});
```

> This approach is costly since the amount of signals sent back on request grows linearly with the amount of connected users. So if there are three signals requested as opposed to one, there are 3 times as many total signals being submitted on every connect. To avoid this costly scenario we can group the signal types into one single signal that captures all the required information:

Reviewer 3: ```suggestion
Each of the N connected clients would then respond with 3 signals as well (3N signals total).  To bring this down to N signals total, we can group these requests into a single request that captures all the required information:
```

```typescript
container.on("connected", () => {
    this.signaler.submitSignal("connectRequest");
});
```

> The idea is that the payload sent back on the `connectRequest` will include all the relevant information the newly connected user needs.

Reviewer 3: ```suggestion
The payload sent back in response to the `connectRequest` should include all the relevant information the newly connected user needs.
```
